File: string/PG_문자열 압축.cpp
#include <string>
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int solution(string s) {
    int answer = 987654321;
    for(double i = 1; i <= s.size()/2+1; i++) {
        vector<string> v;
        string temp = ""; int cnt = 1; string last ="";
        for(int j = 0; j < s.size(); j += i) {
            v.push_back(s.substr(j, i));
        }
        for(int j = 1; j < v.size(); j++) {
            if(v[j] != v[j-1]) {
                if(cnt != 1) {
                    if(cnt >= 10) temp += to_string(cnt); // 숫자가 일의 자리를 넘어가는 경우에 대한 처리가 필요함
                    else temp += (cnt+'0');
                }
                last = v[j-1];
                temp += v[j-1];
                cnt = 1;
            }
            else cnt++;
        }
        if(last != v[v.size()-1]) {
            if(cnt != 1) {
                if(cnt >= 10) temp += to_string(cnt);
                else temp += (cnt+'0');
            }
            temp += v[v.size()-1];
        }
        int ssize = temp.size();
        answer = min(ssize, answer);
    }
    return answer;
}

File: string/PG_문자열 압축.java
// 스터디에서 java로 다시 푼 코드

import java.util.ArrayList;

class Solution {
    public int solution(String s) {
        int answer = 987654321;
        if(s.length()==1) {
            answer = 1;
        }
        for(int i = 1; i <= s.length()/2; i++) {
            String temp = s;
            String ans = "";
            int cnt = 1; int flag = 1;
            ArrayList<String> list = new ArrayList<String>();
            int a;
            for(a = 0; a < s.length()-i; a+=i) {
                if(a+i < s.length()) list.add(temp.substring(a, a+i));
            }
            list.add(temp.substring(a, s.length()));
            for(int j = 0; j < list.size()-1; j++) {
                String t1 = list.get(j);
                String t2 = list.get(j+1);
                if(t1.equals(t2)) {
                    cnt++;
                }
                else {
                    if(j == 0) {
                        ans += list.get(0);
                        continue;
                    }
                    if(cnt != 1) ans += Integer.toString(cnt);
                    ans += list.get(j);
                    cnt = 1;
                }
            }
            if(cnt != 1) ans += Integer.toString(cnt);
            ans += list.get(list.size()-1);
            answer = Math.min(answer, ans.length());
        }
        return answer;
    }
}
